### PR TITLE
CORE: check memtype in collective init

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Do memory type detection if UCC_MEMORY_TYPE_UNKNOWN is set in coll_args.

## Why ?
We want to know memory type as earlier as possible to do CL, TL or algorithm selection. If user doesn't specify memory type then UCC will try to detect it in collective init.

## How ?
Following collective args description in [doc](https://github.com/openucx/ucc/blob/master/docs/images/ucc_coll_args_table1.pdf)
ucc_coll_args_check_mem_type function checks if memory type is unknown and in that case uses mc detect to get memory type.
Questions:
1. The large switch can be collapsed if we move mem_type and data_type from info/info_v to src and dst in coll_args, but it requires changes in ucc_coll_args_t again.
2. For allreduce specification says that in case of inplace src.info.buffer is ignored, but src.info.mem_type and src.info.datatype should be set. It doesn't make sense to set buffer memory type if buffer is ignored, should we change comment to "src.info.buffer, src.info.mem_type, src.info.datatype are ignored"?

### CODE REVIEW

**Review by bureddy — APPROVED** [review]

**Comment by manjugv** [comment]
> 1. The large switch can be collapsed if we move mem_type and data_type from info/info_v to src and dst in coll_args, but it requires changes in ucc_coll_args_t again.

If we do that, we have differentiate between src and dest memtype? how do you want to do that? separate variables?

**Comment by manjugv** [comment]
> 2\. For allreduce specification says that in case of inplace src.info.buffer is ignored, but src.info.mem_type and src.info.datatype should be set. It doesn't make sense to set buffer memory type if buffer is ignored, should we change comment to "src.info.buffer, src.info.mem_type, src.info.datatype are ignored"?

Sounds reasonable. @vspetrov Please comment on that.

**Comment by Sergei-Lebedev** [comment]
> > 1. The large switch can be collapsed if we move mem_type and data_type from info/info_v to src and dst in coll_args, but it requires changes in ucc_coll_args_t again.
> 
> If we do that, we have differentiate between src and dest memtype? how do you want to do that? separate variables?

Sometimes we just need to know datatype or memory type and we don't care if it's v collective or not. So yes, separate variables like src_mem_type, src_data_type, dst_mem_type, dst_data_type will work.
Another way is to permute info and info_v structures fields so that mem_type is first field and datatype is second in both ucc_coll_buffer_info_v_t and ucc_coll_buffer_info_t. I remember in C standard there is a rule about accessing starting common fields of union through one union member or another, but this is less readable in my opinion. WDYT?
@manjugv  @bureddy  @vspetrov

**Comment by vspetrov** [comment]
> > 2. For allreduce specification says that in case of inplace src.info.buffer is ignored, but src.info.mem_type and src.info.datatype should be set. It doesn't make sense to set buffer memory type if buffer is ignored, should we change comment to "src.info.buffer, src.info.mem_type, src.info.datatype are ignored"?
> 
> Sounds reasonable. @vspetrov Please comment on that.

for allreduce: we require setting src.info.buffer,count,data_type,mem_type and dst.info.buffer,mem_type. Note count and datatype are always taken from src (they have to be the same for src and dst). If INPLACE is used then src.info.buffer is ignored, and YES we should ingore src.info.mem_type as well. But count/datatype are still valid in src.info.

**Comment by vspetrov** [comment]
> > > 1. The large switch can be collapsed if we move mem_type and data_type from info/info_v to src and dst in coll_args, but it requires changes in ucc_coll_args_t again.
> > 
> > 
> > If we do that, we have differentiate between src and dest memtype? how do you want to do that? separate variables?
> 
> Sometimes we just need to know datatype or memory type and we don't care if it's v collective or not. So yes, separate variables like src_mem_type, src_data_type, dst_mem_type, dst_data_type will work.
> Another way is to permute info and info_v structures fields so that mem_type is first field and datatype is second in both ucc_coll_buffer_info_v_t and ucc_coll_buffer_info_t. I remember in C standard there is a rule about accessing starting common fields of union through one union member or another, but this is less readable in my opinion. WDYT?
> @manjugv @bureddy @vspetrov

@Sergei-Lebedev we can't eliminate "the switch" entirely, right, just make it smaller? Let try if possible. Will smth like this work:
coll_args {
 src {
      mem_type;
      union {
          info;
           info_v;
      }
 }
dst {
      mem_type;
      union {
           info;
           info_v;
      }
}

Is that what you meant? Another question to everybody (@manjugv @Sergei-Lebedev ) do we want to support multiple datatypes in src/dst: like MPI_Alltoallw (separate datatype for each offset) ? Right now we don't support that. IF so, then i agree with sergey we can also move dtype out of info/info_v

**Comment by Sergei-Lebedev** [comment]
> > > 1. For allreduce specification says that in case of inplace src.info.buffer is ignored, but src.info.mem_type and src.info.datatype should be set. It doesn't make sense to set buffer memory type if buffer is ignored, should we change comment to "src.info.buffer, src.info.mem_type, src.info.datatype are ignored"?
> > 
> > 
> > Sounds reasonable. @vspetrov Please comment on that.
> 
> for allreduce: we require setting src.info.buffer,count,data_type,mem_type and dst.info.buffer,mem_type. Note count and datatype are always taken from src (they have to be the same for src and dst). If INPLACE is used then src.info.buffer is ignored, and YES we should ingore src.info.mem_type as well. But count/datatype are still valid in src.info.

Does it mean that we require to duplicate same value of count and datatype in src.info and dst.info? Why?

**Comment by Sergei-Lebedev** [comment]
> > > > 1. The large switch can be collapsed if we move mem_type and data_type from info/info_v to src and dst in coll_args, but it requires changes in ucc_coll_args_t again.
> > > 
> > > 
> > > If we do that, we have differentiate between src and dest memtype? how do you want to do that? separate variables?
> > 
> > 
> > Sometimes we just need to know datatype or memory type and we don't care if it's v collective or not. So yes, separate variables like src_mem_type, src_data_type, dst_mem_type, dst_data_type will work.
> > Another way is to permute info and info_v structures fields so that mem_type is first field and datatype is second in both ucc_coll_buffer_info_v_t and ucc_coll_buffer_info_t. I remember in C standard there is a rule about accessing starting common fields of union through one union member or another, but this is less readable in my opinion. WDYT?
> > @manjugv @bureddy @vspetrov
> 
> @Sergei-Lebedev we can't eliminate "the switch" entirely, right, just make it smaller? Let try if possible. Will smth like this work:
> coll_args {
> src {
> mem_type;
> union {
> info;
> info_v;
> }
> }
> dst {
> mem_type;
> union {
> info;
> info_v;
> }
> }
> 
> Is that what you meant?
Yes, but we also can move datatype together with memtype since it's common for info and info_v.

**Review by vspetrov — APPROVED** [review]

File: src/core/ucc_coll.c [hunk 1]
```diff
@@ -21,13 +22,108 @@ static ucc_cl_team_t *ucc_select_cl_team(ucc_coll_args_t *coll_args,
...
+        return UCC_OK;
+    case UCC_COLL_TYPE_BCAST:
+        UCC_BUFFER_INFO_CHECK_MEM_TYPE(coll_args->src.info);
+        return UCC_OK;
+    case UCC_COLL_TYPE_ALLREDUCE:
+        UCC_BUFFER_INFO_CHECK_MEM_TYPE(coll_args->dst.info);
+        if (!UCC_IS_INPLACE(*coll_args)) {
+            UCC_BUFFER_INFO_CHECK_MEM_TYPE(coll_args->src.info);
+        } else {
+            coll_args->src.info.mem_type = coll_args->dst.info.mem_type;
```
[the remark this answers is not in the document]
Comment (Sergei-Lebedev) [reply src/core/ucc_coll.c:52]: yes, as we discussed in WG let's make these changes part of next PR

[the remark this answers is not in the document]
Comment (vspetrov) [reply src/core/ucc_coll.c:52]: 👍

**Pull request merged**